misc/python/materialize/output_consistency/generators/expression_generator.py

```diff
@@ -73,46 +72,62 @@ def _initialize_types(self) -> None:
             types_with_values.append(data_type_with_values)
             self.types_with_values_by_category[category] = types_with_values
 
-    def pick_random_operation(self) -> DbOperationOrFunction:
+    def pick_random_operation(self, include_aggregates: bool) -> DbOperationOrFunction:
+        weights = (
+            self.operation_weights
+            if include_aggregates
+            else self.operation_weights_no_aggregates
+        )
+
         return self.randomized_picker.random_operation(
-            self.selectable_operations, self.operation_weights
+            self.selectable_operations, weights
         )
 
     def generate_expression(
         self,
         operation: DbOperationOrFunction,
-        logger: ConsistencyTestLogger,
-        try_number: int = 1,
     ) -> Optional[Expression]:
+        storage_layout = self._select_storage_layout(operation)
         try:
-            args = self.generate_args_for_operation(operation)
+            args = self._generate_args_for_operation(operation, storage_layout)
         except NoSuitableExpressionFound:
             return None
 
+        is_aggregate = operation.is_aggregation or self._contains_aggregate_arg(args)
         is_expect_error = operation.is_expected_to_cause_db_error(args)
-        expression = ExpressionWithArgs(operation, args, is_expect_error)
+        expression = ExpressionWithArgs(operation, args, is_aggregate, is_expect_error)
 
-        if self.known_inconsistencies_filter.matches(expression):
-            if try_number <= 5:
-                if self.config.verbose_output:
-                    logger.add_global_warning(
-                        f"Skipping expression with known inconsistency: {expression}"
-                    )
+        return expression
 
-                return self.generate_expression(
-                    operation, logger, try_number=try_number + 1
-                )
-            else:
-                logger.add_global_warning(
-                    f"Aborting expression generation for {operation},"
-                    f" all tries resulted in known inconsistencies (e.g., {expression})"
-                )
-                return None
+    def _select_storage_layout(
+        self, operation: DbOperationOrFunction
+    ) -> ValueStorageLayout:
+        if not operation.is_aggregation:
+            # Non-aggregate expressions can unfortunately only use horizontal layout because the processing order does
```

Comment:
Do you have an example of this situation?

Reply:
Given the following setup:
```
DROP TABLE IF EXISTS t_dfr_vert;
CREATE TABLE t_dfr_vert (row_index INT, decimal_39_0_val DECIMAL(39));
INSERT INTO t_dfr_vert VALUES (1, 0::DECIMAL(39));
INSERT INTO t_dfr_vert VALUES (9, -999999999999999999999999999999999999999::DECIMAL(39));
-- ------------------------------
-- Setup for evaluation strategy 'Constant folding':
CREATE OR REPLACE VIEW v_ctf_vert (row_index, decimal_39_0_val)
 AS SELECT 1, 0::DECIMAL(39)
    UNION SELECT 9, -999999999999999999999999999999999999999::DECIMAL(39);
```
an example is:
```
-- Query using evaluation strategy 'Dataflow rendering':
SELECT
  ln(decimal_39_0_val)
FROM
  t_dfr_vert
WHERE row_index IN (9, 1)
ORDER BY row_index ASC;
-- ------------------------------
-- Query using evaluation strategy 'Constant folding':
SELECT
  ln(decimal_39_0_val)
FROM
  v_ctf_vert
WHERE row_index IN (9, 1)
ORDER BY row_index ASC;
```

The column `decimal_39_0_val` contains zero and a negative value:
* Data-flow rendering will fail with: `function ln is not defined for negative numbers`
* Constant folding will fail with: `function ln is not defined for zero`

I consider both error messages valid.

Reply:
ok understood.

Reply:
However, this will be a problem when adding the top-k functionality... 😞

Reply:
Yes, I think this is fine, since we cannot guarantee evaluation order under different plans.